internal/locate/region_cache.go

```diff
@@ -2455,12 +2455,20 @@ func (s *Store) reResolve(c *RegionCache) (bool, error) {
 	addr = store.GetAddress()
 	if s.addr != addr || !s.IsSameLabels(store.GetLabels()) {
 		newStore := &Store{storeID: s.storeID, addr: addr, saddr: store.GetStatusAddress(), storeType: storeType, labels: store.GetLabels(), state: uint64(resolved)}
-		newStore.livenessState = atomic.LoadUint32(&s.livenessState)
 		newStore.unreachableSince = s.unreachableSince
 		c.storeMu.Lock()
+		// set new store liveness state in mutex to avoid problem in https://github.com/tikv/client-go/issues/1401
+		newStore.livenessState = atomic.LoadUint32(&s.livenessState)
```

Comment:
Another way to fix it is always set newStore's livenessState to reachable.

Reply:
Aways setting the newStore to reachable may lead to issues like https://github.com/tikv/client-go/issues/1111 .

It seems that this PR cannot handle the case of changing of the store addr.

The root cause of the bug is that an unreachable new store can be inserted without starting health check, because the reResovle can be called by health-check-loop and check-and-resolve-loop and only health-check-loop will start health check for new store when it's unreachable. So, how about to start health check in reResolve when a unreachable new store inserted?

Reply:
LGTM, except for that the code of triggering health check in health-check-loop need to be removed.

Reply:
Besides, we may also check the resolveState in health check loop, return if the current store is deleted.

Reply:
done, PTAL